cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java

```diff
@@ -79,11 +84,79 @@ public AggregatedMetricsEmitter load(EmitterKey key) throws Exception {
    */
   protected abstract void publish(Iterator<MetricValue> metrics) throws Exception;
 
+  /**
+   * @return true if we want to publish metrics about the received metrics in {@link #publish(java.util.Iterator)}.
+   */
+  protected boolean isPublishMetaMetrics() {
+    return true;
+  }
+
   @Override
   protected final void runOneIteration() throws Exception {
     final long timestamp = TimeUnit.SECONDS.convert(System.currentTimeMillis(), TimeUnit.MILLISECONDS);
     LOG.trace("Start log collection for timestamp {}", timestamp);
-    Iterator<MetricValue> metricsItor = getMetrics(timestamp);
+
+    final Iterator<MetricValue> rawMetricsItor = getMetrics(timestamp);
+    Iterator<MetricValue> metricsItor;
+    if (!isPublishMetaMetrics()) {
+      metricsItor = rawMetricsItor;
+    } else {
+      // wrap the raw metrics iterator with an iterator that will publish meta metrics
+      metricsItor = new AbstractIterator<MetricValue>() {
+        private Queue<MetricValue> pendingMetrics = new LinkedList<MetricValue>();
+        private boolean queuedAggregateMetrics = false;
+        private int numMetrics = 0;
+
+        @Override
+        protected MetricValue computeNext() {
+          if (rawMetricsItor.hasNext()) {
+
+            if (pendingMetrics.isEmpty()) {
+              // publish raw metric
+              MetricValue raw = rawMetricsItor.next();
+              numMetrics++;
+
+              // sample processing delay - assumes we mostly receive metrics in time order
+              if (raw.getTimestamp() - timestampLastSampled >= SAMPLING_DELAY_SEC) {
```

Comment:
should we also have another option to execute this (every 'n' seconds ?) , let's say we have ```SAMPLING_DELAY_SEC= 10s``` , and we emit 10k records at ```timestamp -1428023632``` and let's say it took 20seconds (hypothetically) to process, assuming we are continuing process at the same rate, by the time we reach ```timestamp -1428023732``` and execute the next iteration, the latency would be roughly close to 3 minutes, we might want to get this latency information more frequently right ?